File: sdk/python/examples/09_camera_stream.py
"""MJPEG 流服务器示例：让远程浏览器查看摄像头画面。

用法:
    # 默认端口 8080
    python 09_camera_stream.py

    # 自定义端口
    python 09_camera_stream.py 9090

启动后，在浏览器中打开:
    http://<机器狗IP>:8080/           — 带样式的查看页面
    http://<机器狗IP>:8080/stream     — 原始 MJPEG 流
    http://<机器狗IP>:8080/snapshot   — 单帧 JPEG 快照
"""

import sys
import signal

from brainstem_sdk.camera import OrixCamera

# ── 参数解析 ────────────────────────────────────────────────

port = 8080
source = 0

for arg in sys.argv[1:]:
    if arg.isdigit():
        port = int(arg)
    else:
        source = arg

# ── 启动流服务器 ────────────────────────────────────────────

cam = OrixCamera(source=source)

if not cam.open():
    print("无法打开摄像头，请检查设备连接。")
    sys.exit(1)

w, h = cam.get_frame_size()
print(f"摄像头已打开: {w}x{h}")


def shutdown(signum, frame):
    print("\n正在关闭...")
    cam.close()
    sys.exit(0)


signal.signal(signal.SIGINT, shutdown)
signal.signal(signal.SIGTERM, shutdown)

cam.stream_mjpeg(port=port)

print(f"MJPEG 流服务器已启动于端口 {port}")
print(f"  查看页面: http://0.0.0.0:{port}/")
print(f"  原始流:   http://0.0.0.0:{port}/stream")
print(f"  快照:     http://0.0.0.0:{port}/snapshot")
print()
print("按 Ctrl+C 停止服务器")

# 保持主线程运行
try:
    signal.pause()
except AttributeError:
    # Windows 不支持 signal.pause()，用循环替代
    import time
    while True:
        time.sleep(1.0)

File: sdk/cpp/examples/09_camera_stream.cpp
// 09_camera_stream.cpp — MJPEG HTTP 流服务器
//
// 启动后，在浏览器访问：
//   http://<机器人IP>:8080/stream    — 实时视频流
//   http://<机器人IP>:8080/snapshot  — 单帧快照
//
// 按 Ctrl+C 停止。
#include "orix/orix_camera.h"

#include <chrono>
#include <csignal>
#include <iostream>
#include <thread>

static volatile bool g_interrupt = false;
static void sig_handler(int) { g_interrupt = true; }

int main(int argc, char* argv[]) {
    int port = 8080;
    int src  = 0;
    if (argc >= 2) src  = std::stoi(argv[1]);
    if (argc >= 3) port = std::stoi(argv[2]);

    std::signal(SIGINT, sig_handler);

    try {
        orix::OrixCamera cam(src, 640, 480, 30);
        cam.open();

        auto [w, h] = cam.get_frame_size();
        std::cout << "摄像头: " << w << "x" << h << "\n";

        std::cout << "启动 MJPEG 流服务器 (端口 " << port << ")...\n";
        cam.stream_mjpeg(port, "0.0.0.0");

        std::cout << "已启动！浏览器访问：\n"
                  << "  http://localhost:" << port << "/stream    （实时流）\n"
                  << "  http://localhost:" << port << "/snapshot  （快照）\n"
                  << "按 Ctrl+C 停止。\n\n";

        // 主线程等待，每秒打印一次状态
        long secs = 0;
        while (!g_interrupt) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
            ++secs;
            std::cout << "\r运行中 " << secs << "s..." << std::flush;
        }

        std::cout << "\n停止流服务器...\n";
        cam.stop_stream();
        cam.close();
        std::cout << "已停止。\n";

    } catch (const orix::CameraOpenError& e) {
        std::cerr << "摄像头错误: " << e.what() << "\n";
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "错误: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
